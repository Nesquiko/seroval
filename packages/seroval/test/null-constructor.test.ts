import { describe, it, expect } from 'vitest';
import {
  compileJSON,
  deserialize,
  Feature,
  fromJSON,
  serialize,
  serializeAsync,
  toJSON,
  toJSONAsync,
} from '../src';

describe('null-constructor', () => {
  describe('serialize', () => {
    it('supports Object.create(null)', () => {
      const example = Object.assign(Object.create(null), { hello: 'world' }) as { hello: string };
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Record<string, string>>(result);
      expect(back.constructor).toBeUndefined();
      expect(back.hello).toBe(example.hello);
    });
  });
  describe('serializeAsync', () => {
    it('supports Object.create(null)', async () => {
      const example = Object.assign(Object.create(null), { hello: 'world' }) as { hello: string };
      const result = await serializeAsync(example);
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<Record<string, string>>>(result);
      expect(back.constructor).toBeUndefined();
      expect(back.hello).toBe(example.hello);
    });
  });
  describe('toJSON', () => {
    it('supports Object.create(null)', () => {
      const example = Object.assign(Object.create(null), { hello: 'world' }) as { hello: string };
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Record<string, string>>(result);
      expect(back.constructor).toBeUndefined();
      expect(back.hello).toBe(example.hello);
    });
  });
  describe('toJSONAsync', () => {
    it('supports Object.create(null)', async () => {
      const example = Object.assign(Object.create(null), { hello: 'world' }) as { hello: string };
      const result = await toJSONAsync(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<Record<string, string>>>(result);
      expect(back.constructor).toBeUndefined();
      expect(back.hello).toBe(example.hello);
    });
  });
  describe('compat', () => {
    it('should use manual assignment instead of Object.assign', () => {
      const example = Object.assign(Object.create(null), { hello: 'world' }) as { hello: string };
      const result = serialize(example, {
        disabledFeatures: Feature.ObjectAssign,
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<Record<string, string>>(result);
      expect(back.constructor).toBeUndefined();
      expect(back.hello).toBe(example.hello);
    });
  });
  describe('compat#toJSON', () => {
    it('should use manual assignment instead of Object.assign', () => {
      const example = Object.assign(Object.create(null), { hello: 'world' }) as { hello: string };
      const result = toJSON(example, {
        disabledFeatures: Feature.ObjectAssign,
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      expect(compileJSON(result)).toMatchSnapshot();
      const back = fromJSON<Record<string, string>>(result);
      expect(back.constructor).toBeUndefined();
      expect(back.hello).toBe(example.hello);
    });
  });
});
