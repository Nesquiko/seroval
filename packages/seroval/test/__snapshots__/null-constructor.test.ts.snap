// Vitest Snapshot v1

exports[`null-constructor > compat > should use manual assignment instead of Object.assign 1`] = `"(h=>((h=Object.create(null),h.hello=\\"world\\",h)))()"`;

exports[`null-constructor > compat#toJSON > should use manual assignment instead of Object.assign 1`] = `"{\\"t\\":{\\"t\\":17,\\"i\\":0,\\"d\\":{\\"k\\":[\\"hello\\"],\\"v\\":[{\\"t\\":1,\\"s\\":\\"world\\"}],\\"s\\":1}},\\"r\\":0,\\"i\\":false,\\"f\\":8063,\\"m\\":[]}"`;

exports[`null-constructor > compat#toJSON > should use manual assignment instead of Object.assign 2`] = `"(h=>((h=Object.create(null),h.hello=\\"world\\",h)))()"`;

exports[`null-constructor > serialize > supports Object.create(null) 1`] = `"Object.assign(Object.create(null),{hello:\\"world\\"})"`;

exports[`null-constructor > serializeAsync > supports Object.create(null) 1`] = `"Object.assign(Object.create(null),{hello:\\"world\\"})"`;

exports[`null-constructor > toJSON > supports Object.create(null) 1`] = `"{\\"t\\":{\\"t\\":17,\\"i\\":0,\\"d\\":{\\"k\\":[\\"hello\\"],\\"v\\":[{\\"t\\":1,\\"s\\":\\"world\\"}],\\"s\\":1}},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;

exports[`null-constructor > toJSONAsync > supports Object.create(null) 1`] = `"{\\"t\\":{\\"t\\":17,\\"i\\":0,\\"d\\":{\\"k\\":[\\"hello\\"],\\"v\\":[{\\"t\\":1,\\"s\\":\\"world\\"}],\\"s\\":1}},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;
